### PR TITLE
Prevents null pointers when doing consecutive VM migrates

### PR DESCRIPTION
**Pull request by RodrigoDLopez**
## Description

If a VM gets consecutive migrations, a null pointer exception is thrown because these VMs do not have `host_id` or` last_host_id`; ACS clears these fields when the first migration is over.
With that in mind, this PR prevents the respective null pointer. Additionally, it logs the right context and gives some information to the operator.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

To test this, I used the `cloudmonkey` to request `migrateVirtualMachine` commands

### CODE REVIEW

**Comment by RodrigoDLopez** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Review by DaanHoogland — APPROVED** [review]
changes look good, starting tests on assorted platforms as I'm not sure this matters

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -2372,18 +2368,32 @@ private void setDestinationPoolAndReallocateNetwork(StoragePool destPool, VMInst
...
+            return;
+        }
+        Long srcClusterId = srcHost.getClusterId();
+        if (srcClusterId.equals(destClusterId)) {
+            s_logger.debug("Since the Source cluster ID [%s] is equal to the Destination cluster ID [%s] we do not need to proceed with the clean up after migration");
+            return;
+        }
+        String srcDcName = _clusterDetailsDao.getVmwareDcName(srcClusterId);
+        String destDcName = _clusterDetailsDao.getVmwareDcName(destClusterId);
+        if(StringUtils.isNotBlank(srcDcName) && StringUtils.isNotBlank(destDcName)){
```
Comment (rhtyd) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:2391]: @RodrigoDLopez shouldn't we be doing a `!srcDcName.equals(destDcName)` check here?

**Comment by ravening** [comment]
@RodrigoDLopez I tried migrating same vm like 10 times between 5 hypervisors and never got any NPE. Also the `lost_host_id` field was not cleared.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 1 above)
Comment (RodrigoDLopez) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:2391]: Hi @rhtyd, on line 2385 we check if the srcClusterId is equal to destClusterId and if so, ACS will show a log message give information about this and do nothing

**Comment by RodrigoDLopez** [comment]
Hi, @ravening can you report the methodology you used?
* ACS version
* How are you executing the tests? Are you migrating VMs running? Stopped VMs? Stopped VMs, and then starting them, and just then migrating again?

If you take a look at the code, a null pointer will for sure happen. Just look at the lines I am changing.

Before this PR, the first migration will have `host_id` or `last_host_id`, but during the migrate, ACS will set the field `last_host_id` to null using the `afterStorageMigrationCleanup` method, and on the next executions of the migrate command (without starting the VM), this field will be null and then on this line, a NPE will be launched
https://github.com/apache/cloudstack/blob/d657ef7d5b4594b638eaf7df312b11efb99000fb/engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java#L2282-L2283

Furthermore, some of these parameters were never used for some hypervisors, seems like VMWare needs it, that is why I moved it where they will be used, and did the necessary validations providing information about the process with an appropriate log message

**Comment by ravening** [comment]
> Hi, @ravening can you report the methodology you used?
> 
> * ACS version
> 
> * How are you executing the tests? Are you migrating VMs running? Stopped VMs? Stopped VMs, and then starting them, and just then migrating again?
> 
> 
> 
> If you take a look at the code, a null pointer will for sure happen. Just look at the lines I am changing.
> 
> 
> 
> Before this PR, the first migration will have `host_id` or `last_host_id`, but during the migrate, ACS will set the field `last_host_id` to null using the `afterStorageMigrationCleanup` method, and on the next executions of the migrate command (without starting the VM), this field will be null and then on this line, a NPE will be launched
> 
> https://github.com/apache/cloudstack/blob/d657ef7d5b4594b638eaf7df312b11efb99000fb/engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java#L2282-L2283
> 
> 
> 
> Furthermore, some of these parameters were never used for some hypervisors, seems like VMWare needs it, that is why I moved it where they will be used, and did the necessary validations providing information about the process with an appropriate log message

@RodrigoDLopez im using ACS 4.14

I doing regular VM migration between multiple hosts. I didn't do migration with storage. Also I'm doing migration on KVM host as not VMware

**Comment by RodrigoDLopez** [comment]
Hi @ravening, that was my fault.
You are doing a live migration, this NPE will occur when you: 
- stop an instance;
- migrate it with volume
- do not start the instance
- try to migrate again to another pool

with these steps, after the first migrate `host_id` and `last_host_id` will be set to `null`. And then, when you try to migrate it again a NPE will be thrown.

**Comment by ravening** [comment]
> Hi @ravening, that was my fault.
> 
> You are doing a live migration, this NPE will occur when you: 
> 
> - stop an instance;
> 
> - migrate it with volume
> 
> - do not start the instance
> 
> - try to migrate again to another pool
> 
> 
> 
> with these steps, after the first migrate `host_id` and `last_host_id` will be set to `null`. And then, when you try to migrate it again a NPE will be thrown.

@RodrigoDLopez ah ok... So you are just doing volume migration of VM...I will try these steps and let you know the result

**Comment by RodrigoDLopez** [comment]
@rhtyd, @ravenin, @DaanHoogland 

Hello guys, happy new year.

Is there something missing here? Everything seems to be ok with the patch. It is a good bug fix, it has no errors, and tests are passing

**Comment by RodrigoDLopez** [comment]
@ravening @rhtyd @harikrishna-patnala @nvazquez 

Hello everyone, this PR fixes a bug and is a simple change.  
I believe that it is in the interest of the whole community that this bug found to be solved.
would it be possible to receive some reviews or tests in this PR?

**Comment by ravening** [comment]
> @ravening @rhtyd @harikrishna-patnala @nvazquez 
> 
> 
> 
> Hello everyone, this PR fixes a bug and is a simple change.  
> 
> I believe that it is in the interest of the whole community that this bug found to be solved.
> 
> would it be possible to receive some reviews or tests in this PR?

Will review and test it this week

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 2]
```diff
@@ -2372,18 +2368,32 @@ private void setDestinationPoolAndReallocateNetwork(StoragePool destPool, VMInst
...
+            s_logger.debug(String.format("No Host ID was found when doing cleanup after Vmware migration for VM with ID = [%d] and Cluster destination ID = [%d]", vm.getId(), destClusterId));
+            return;
+        }
+        HostVO srcHost = _hostDao.findById(srcHostId);
+        if (srcHost == null) {
+            s_logger.debug(String.format("When doing cleanup after Vmware migration could not find a host for the given ID = [%d]", srcHostId));
+            return;
+        }
+        Long srcClusterId = srcHost.getClusterId();
+        if (srcClusterId.equals(destClusterId)) {
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:2385]: Suggestion: why not reversing the logic and moving the code block from in lines 2377-2383 to this point?

I'm ok anyways but I think the other way is easier to read. If you are keeping this if block please add null checks for srcClusterId and destClusterId

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 1 above)
Comment (nvazquez) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:2391]: Minor checkstyle bug: please add a space between the if and the condition and other space before the curly brace after the condition

**Review by nvazquez — COMMENTED** [review]
Looks good, I've left a few comments

**Comment by RodrigoDLopez** [comment]
> Looks good, I've left a few comments

I will make this changes. Thanks for the hint

**Comment by RodrigoDLopez** [comment]
> @RodrigoDLopez do you want this on latest or on one of the release branches (4.14 or 4.15)?

For me, it's fine to be merged just on the master, since it's a nonreported bug. But if you wanna I can do the necessary changes to merge this fix on 4.14 as well

@nvazquez @ravening @rhtyd @harikrishna-patnala 
any update related to this PR? something that maybe I need to change?

**Comment by rhtyd** [comment]
@RodrigoDLopez can you fix the conflict?

**Comment by GabrielBrascher** [comment]
I just reproduced this very issue, this PR would be nice to have (4.15.2+, I think hat for 4.15.1 it is not feasible).
@RodrigoDLopez can you please fix the conflicts?

**Comment by shwstppr** [comment]
@GabrielBrascher were you able to reproduce this with 4.15.1 RC?
From the code changes, I feel this affects only VMware. Is that correct @RodrigoDLopez ?
Last time I was not able to reproduce this after #4385. Can check again

**Comment by GabrielBrascher** [comment]
@rhtyd we got it on a 4.15.0.0, but it seems that the codebase at 4.15.1.0 RC1 did not change at that point.
@shwstppr I think that it can happen with any hypervisor as long as one migrate the VM, I saw this issue with KVM.

Steps to reproduce:
~~~
Case: migrate VM (e.g. for maintenance), keep VM stopped, try to migrate back the VM -> NullPointer

1. Migrate VM from host A to host B
2. VM stills stopped at host B
3. Try to migrate VM back to host A
4. Null pointer
~~~

Stack trace:
```
java.lang.NullPointerException
        at com.cloud.vm.VirtualMachineManagerImpl.migrateThroughHypervisorOrStorage(VirtualMachineManagerImpl.java:2302)
        at com.cloud.vm.VirtualMachineManagerImpl.orchestrateStorageMigration(VirtualMachineManagerImpl.java:2194)
        at com.cloud.vm.VirtualMachineManagerImpl.orchestrateStorageMigration(VirtualMachineManagerImpl.java:5625)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```

When checking the stack trace + DB it is possible to detect that the issue is indeed caused by the fact of last `last_host_id` being null, as reported by @RodrigoDLopez.

**Comment by shwstppr** [comment]
@GabrielBrascher I've not tested yet and will try it tomorrow but I feel there has been a change in code that would prevent NPE that you shared,
4.15.0.0 (failing to get clusterId for the source host as srcHost is null): https://github.com/apache/cloudstack/blob/4.15.0.0/engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java#L2302
4.15 branch (logic for finding source clusterId has been changed. It is now found from `host_id` or `last_host_id` or storage pool of the VMs volumes): https://github.com/apache/cloudstack/blob/4.15/engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java#L2318

**Comment by DaanHoogland** [comment]
@RodrigoDLopez can you address the conflict?
@rhtyd @nvazquez are your comments addressed to satisfaction?

**Comment by GabrielBrascher** [comment]
@shwstppr thanks for bringing those changes, I might have tested with 4.15.0.0 then instead of RC1. I will take another look at it.

**Comment by GabrielBrascher** [comment]
@rhtyd @RodrigoDLopez @shwstppr @DaanHoogland I have been running some tests on RC2 and I was not able to reproduce this issue with 4.15.1.0 RC2.

This one looks to be fixed.

**Comment by rhtyd** [comment]
Thanks for testing @GabrielBrascher, @RodrigoDLopez can you test 4.15.1.0 RC2?

**Comment by RodrigoDLopez** [comment]
not needed anymore...

**Comment by rafaelweingartner** [comment]
@RodrigoDLopez why is this one not needed anymore? Was it addressed somewhere else?

**Pull request closed without merging**